src/models/karaoke.py

```diff
@@ -0,0 +1,220 @@
+    karaStart: timedelta() = timedelta()
+    karaEnd: timedelta() = timedelta()
+    karaDuration: timedelta() = timedelta()
+
+    line: KLine = None
+    syl: KSyl = None
+
+
+@dataclass
+class KSyl:
+    start: timedelta()
+    end: timedelta()
+    duration: timedelta()
+    chars: Sequence[KChar]
+    text: str
+    inlineFx: str
+    i: int
+
+    line: KLine = None
+
+    def calculate_char_kara_times(self, style: Style):
+        fontScaler = FontScaler(style.fontName, style.fontSize)
+
+        syllableCharLengths = [
+            timedelta(milliseconds=c * 10)
+            for c in fontScaler.split_by_rendered_width(
+                pyass.timedelta(self.duration).total_centiseconds(), self.text
+            )
+        ]
+
+        sylAccLength = self.start
+
+        for char, syllableCharLength in zip(self.chars, syllableCharLengths):
+            char.karaStart = sylAccLength
+            char.karaDuration = syllableCharLength
+            char.karaEnd = sylAccLength + syllableCharLength
+
+            sylAccLength += syllableCharLength
+
+
+@dataclass
+class KLine:
+    start: timedelta()
+    end: timedelta()
+    duration: timedelta()
+    kara: Sequence[KSyl]
+    text: str
+    startActor: str
+    actorSwitches: Sequence[tuple[int, str]]
+    isSecondary: bool
+    isAlone: bool
+    isEN: bool
+    lineNum: int
+
+    @property
+    def chars(self) -> Sequence[KChar]:
+        return [char for k in self.kara for char in k.chars]
+
+    def calculate_char_offsets(self, style: Style, transitionDuration: timedelta):
+        self.calculate_char_fade_offsets(style, transitionDuration)
+        for k in self.kara:
+            k.calculate_char_kara_times(style)
+
+    def calculate_char_fade_offsets(self, style: Style, transitionDuration: timedelta):
+        fontScaler = FontScaler(style.fontName, style.fontSize)
+
+        lineCharTimes = [
+            timedelta(milliseconds=m)
+            for m in fontScaler.split_by_rendered_width(
+                pyass.timedelta(transitionDuration).total_milliseconds(), self.text
+            )
+        ]
+
+        charAccTime = timedelta()
+        for char, lineCharTime in zip(self.chars, lineCharTimes):
+            char.fadeOffset = charAccTime
+            charAccTime += lineCharTime
+
+
+def preproc_line_text(line: SongLine, lineNum: int = 0) -> KLine:
+    timedeltaUpToIdx = reduce(
+        lambda a, b: a + [a[-1] + b.length], line.syllables, [timedelta(0)]
+    )
+
+    kLine = KLine(
+        start=line.start,
+        end=line.end,
+        duration=line.end - line.start,
+        kara=[],
+        text=line.romaji,
+        startActor=line.actors[0],
+        actorSwitches=[
+            (timedeltaUpToIdx[breakpoint], actor)
+            for breakpoint, actor in zip(line.breakpoints, line.actors)
+            if breakpoint != 0
+        ],
+        isSecondary=line.isSecondary,
+        isAlone=line.romaji == line.en,
+        isEN=False,
+        lineNum=lineNum,
```

Comment:
Would a line ever need to know its current index in the song?

Reply:
Yeah, if someone (well realistically, me...) wants to have different effects for certain lines. It can function as a replacement for the current jank system of changing the line style.

Reply:
I'm not a fan of including a dependency on the exact index of the line within a song to be honest.

Even though realistically it may not change that much I still think it's cleaner to keep all information that relates to a specific line within the line itself (whether as a style/effect/tag/some other field) and at least try to keep each line independent (unless there exists effects that requires information from multiple song lines)

Reply:
In what way? The exact index of the line is only relevant for this script and isn't output into the .ass file, and can be used by any effect generator scripts as a custom modifier for each song to treat certain lines in a special way (yousoroad, kananrail etc.) I don't like the current method where any extra effects for certain special lines needs to be given a special style and then the karaoke templater run for those styles because debugging karaoke templater scripts is a pain.

I guess an alternative is to keep the current karaoke modifier info in the sheets and load it in as a field into the `KLine`, but if we do that, then I think it makes more sense to fold _all_ modifier data into a dictionary within `KLine` instead, and leave the karaoke generation script(s) to handle the modifiers as each one sees fit.

Reply:
I guess the contention here is what we define an "effect" to be.

If an "effect" is "something you apply to a single line" (which is supported by the signature of `to_*_event(KLine) -> pyass.Event`), the decision of which lines to apply the transform function to should lie outside the function, in that the caller of the function should be the one who decides which "effect" to apply to what lines in a song.

In an "effect" is "something you apply to a song" then the decision of which lines to apply particular transformations to is internal to the effect, which I think you're advocating for here.

I don't think the two ideas are mutually incompatible. Perhaps we could have something like:
```python
def to_*_events(song: Song) -> list[pyass.Events]:
    # Determine, using the information in Song, which effect to apply to which lines
    # For example, only apply a specific effect to the bridge of the song
    for line in song
        if /* something */:
            event = to_some_event(to_k_line(line))
        else:
            event = to_some_other_event(to_k_line(line))

def to_some_event(KLine) -> pyass.Event:
    # Generate an pyass.Event based on info in KLine alone
```

The consequence of this is that we can keep the two decisions of "what transformation should we apply to each line of the song?" and "how do we transform a line?" separate, which allows for reusing transformations (`to_*_event`) functions in different songs if necessary

Reply:
I'll be honest, I have the former in right now just because that's what the current `to_events()` had, and I didn't want to change it just yet. But I think that the latter makes more sense for songs in general, since special effects are tied to songs instead of certain lines (and special effects for certain lines aren't really going to be reused between songs 99% of the time).

The question lies in that _something_, is that going to be line numbers? Or is it going to be something like modifier text for certain lines?

Reply:
You're right that in the end there has to be _some_ way to determine which transformation functions to call on which lines of which song, and for this project to continue working this _something_ has to either be in the sheets itself or as additional meta lines in the .ass file (since those are the only two inputs available).

I guess it's going to be hard to come to a good conclusion now given that there's no example usage of the thing yet.

We can just leave it in for now and think about it a bit more when we actually have an effect coded in, then we can revisit this after the dynamic dispatch thing is set up.

Reply:
Sounds good then.